File: frontend/src/services/locations.ts
import { api } from './api';

// A-05: 地点评分/评价/地点知识层（backend/app/api/locations.py 契约对齐）

// 地点列表/详情响应
export interface LocationItem {
  id: number;
  school_id: number;
  name: string;
  description?: string | null;
  latitude: number;
  longitude: number;
  floor?: string | null;
  building?: string | null;
  post_count: number;
  is_verified: boolean;
  // REV-01: 评分汇总
  avg_score: number;
  rating_count: number;
  review_count: number;
}

export interface LocationFactItem {
  id: number;
  location_id: number;
  fact_key: string;
  label: string;
  value: string;
  sort_order: number;
  source_note?: string | null;
  approved_at?: string | null;
  updated_at: string;
}

export interface LocationSummarySource {
  source_type: 'post' | 'review' | 'fact';
  source_id: number;
  title?: string | null;
  snippet?: string | null;
  created_at?: string | null;
  author_name?: string | null;
  score?: number | null;
  confirmation_count?: number;
  refutation_count?: number;
}

export interface LocationSummaryClaim {
  claim_id: string;
  text: string;
  confidence_level: string;
  source_refs: Array<{ source_type: string; source_id: number }>;
}

export interface LocationSummary {
  id?: number | null;
  version?: number | null;
  status: string;
  summary_text?: string | null;
  confidence_level: string;
  claims: LocationSummaryClaim[];
  conflicts: Array<{ text: string; source_refs: Array<{ source_type: string; source_id: number }> }>;
  source_count: number;
  generated_at?: string | null;
  stale_at?: string | null;
  sources: LocationSummarySource[];
}

export interface ReviewAuthor {
  id: number;
  nickname: string;
  avatar_url?: string | null;
  is_verified?: boolean;
}

export interface LocationReviewItem {
  id: number;
  location_id: number;
  user_id: number;
  score: number;
  content?: string | null;
  created_at: string;
  updated_at: string;
  author?: ReviewAuthor | null;
}

export interface LocationDetail {
  location: LocationItem;
  my_review?: LocationReviewItem | null;
  facts: LocationFactItem[];
  summary: LocationSummary;
}

export interface PaginatedReviews {
  items: LocationReviewItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface ReviewPayload {
  score: number;
  content?: string;
}

export interface LocationFactProposalPayload {
  upserts?: Array<{
    fact_key: string;
    label?: string;
    value: string;
    sort_order?: number;
    source_note?: string;
  }>;
  remove_keys?: string[];
  reason?: string;
}

export const locationsApi = {
  /** 当前学校全部地点（含评分汇总），供地图/地点页使用 */
  getLocations: async (): Promise<LocationItem[]> => {
    const response = await api.get('/locations');
    return response.data;
  },

  /** 地点详情（含评分汇总 + 我的评价） */
  getDetail: async (locationId: number): Promise<LocationDetail> => {
    const response = await api.get(`/locations/${locationId}`);
    return response.data;
  },

  /** 地点评价列表 */
  getReviews: async (
    locationId: number,
    page = 1,
    pageSize = 20
  ): Promise<PaginatedReviews> => {
    const response = await api.get(`/locations/${locationId}/reviews`, {
      params: { page, page_size: pageSize },
    });
    return response.data;
  },

  /** 提交/更新本人评价（每地点每用户一条，重复提交=更新） */
  submitReview: async (
    locationId: number,
    payload: ReviewPayload
  ): Promise<LocationReviewItem> => {
    const response = await api.post(`/locations/${locationId}/reviews`, payload);
    return response.data;
  },

  /** 提交地点稳定资料提议（仅校园认证用户） */
  submitFactProposal: async (
    locationId: number,
    payload: LocationFactProposalPayload,
  ): Promise<{ id: number; status: string }> => {
    const response = await api.post(`/locations/${locationId}/fact-proposals`, payload);
    return response.data;
  },

  /** 查看地点 AI 摘要与可追溯来源 */
  getSummary: async (locationId: number): Promise<LocationSummary> => {
    const response = await api.get(`/locations/${locationId}/summary`);
    return response.data;
  },

  /** 撤回本人评价 */
  withdrawReview: async (locationId: number): Promise<{ message: string }> => {
    const response = await api.delete(`/locations/${locationId}/reviews`);
    return response.data;
  },
};

File: backend/app/api/locations.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from sqlalchemy.orm import joinedload
from typing import Optional
from datetime import datetime

from app.database import get_db
from app.dependencies import get_current_user, get_current_user_optional
from app.models.location import Location
from app.models.location_review import LocationReview
from app.models.user import User
from app.schemas.location_review import (
    LocationResponse,
    LocationReviewCreate,
    LocationReviewResponse,
    LocationDetailResponse,
)
from app.schemas.common import PaginatedResponse
from app.schemas.post import UserBrief
from app.core.exceptions import NotFoundException, ForbiddenException, BadRequestException
from app.core.tenant import TenantContext, get_tenant_context, check_resource_in_tenant
from app.core.permissions import require_campus_verified
from app.services.location_summary import (
    load_current_summary,
    load_location_facts,
    load_summary_sources,
    summary_response,
    mark_location_summary_dirty,
)

router = APIRouter(tags=["地点"])


def _location_response(location: Location) -> LocationResponse:
    return LocationResponse(
        id=location.id,
        school_id=location.school_id,
        name=location.name,
        description=location.description,
        latitude=float(location.latitude),
        longitude=float(location.longitude),
        floor=location.floor,
        building=location.building,
        post_count=location.post_count,
        is_verified=location.is_verified,
        avg_score=float(location.avg_score or 0),
        rating_count=location.rating_count,
        review_count=location.review_count,
    )


def _review_response(review: LocationReview) -> LocationReviewResponse:
    author = None
    if review.user:
        # campus_verified 字段由工作流 B（校园认证）新增；此处防御性读取
        if review.is_anonymous:
            # UC-01: 用户离校后评价匿名化——作者显示「已离校用户」，移除认证徽标
            author = UserBrief(
                id=review.user.id,
                nickname="已离校用户",
                avatar_url=None,
                is_verified=False,
            )
        else:
            author = UserBrief(
                id=review.user.id,
                nickname=review.user.nickname,
                avatar_url=review.user.avatar_url,
                is_verified=bool(getattr(review.user, "campus_verified", False)),
            )
    return LocationReviewResponse(
        id=review.id,
        location_id=review.location_id,
        user_id=review.user_id,
        score=review.score,
        content=review.content,
        created_at=review.created_at,
        updated_at=review.updated_at,
        author=author,
    )


async def _recalc_location_rating(db: AsyncSession, location: Location) -> None:
    """REV-01: 重算地点评分汇总（应用服务为唯一写入口，不引入触发器）。"""
    row = (
        await db.execute(
            select(
                func.count(LocationReview.id),
                func.coalesce(func.avg(LocationReview.score), 0),
            ).where(
                LocationReview.location_id == location.id,
                LocationReview.is_deleted == False,
            )
        )
    ).one()
    count, avg = row
    location.review_count = int(count)
    location.rating_count = int(count)
    location.avg_score = round(float(avg), 2)


@router.get("/locations/{location_id}", response_model=LocationDetailResponse, summary="地点详情")
async def get_location(
    location_id: int,
    current_user: Optional[User] = Depends(get_current_user_optional),
    db: AsyncSession = Depends(get_db),
    tenant: TenantContext = Depends(get_tenant_context),
):
    """地点详情（含评分汇总 + 我的评价）。跨校返回 404。"""
    result = await db.execute(
        select(Location).where(Location.id == location_id, Location.is_deleted == False)
    )
    location = result.scalar_one_or_none()
    if location is None:
        raise NotFoundException(detail="地点不存在")
    check_resource_in_tenant(location.school_id, tenant)

    my_review = None
    if current_user is not None:
        rr = await db.execute(
            select(LocationReview).options(joinedload(LocationReview.user)).where(
                LocationReview.location_id == location_id,
                LocationReview.user_id == current_user.id,
                LocationReview.is_deleted == False,
            )
        )
        review = rr.scalar_one_or_none()
        if review is not None:
            my_review = _review_response(review)

    facts = await load_location_facts(db, location.id, tenant.school_id)
    current_summary = await load_current_summary(db, location)
    sources = await load_summary_sources(db, current_summary, tenant) if current_summary else []
    return LocationDetailResponse(
        location=_location_response(location),
        my_review=my_review,
        facts=facts,
        summary=summary_response(current_summary, sources),
    )


@router.get("/locations/{location_id}/reviews", response_model=PaginatedResponse[LocationReviewResponse], summary="地点评价列表")
async def list_location_reviews(
    location_id: int,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    tenant: TenantContext = Depends(get_tenant_context),
):
    # 校验地点存在且属于当前学校
    lr = await db.execute(
        select(Location).where(Location.id == location_id, Location.is_deleted == False)
    )
    location = lr.scalar_one_or_none()
    if location is None:
        raise NotFoundException(detail="地点不存在")
    check_resource_in_tenant(location.school_id, tenant)

    base = select(LocationReview).where(
        LocationReview.location_id == location_id,
        LocationReview.is_deleted == False,
    )
    total = (await db.execute(select(func.count()).select_from(base.subquery()))).scalar() or 0

    query = (
        base.options(joinedload(LocationReview.user))
        .order_by(LocationReview.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    )
    reviews = (await db.execute(query)).scalars().all()
    items = [_review_response(r) for r in reviews]
    return PaginatedResponse.create(items=items, page=page, page_size=page_size, total=total)


@router.post("/locations/{location_id}/reviews", response_model=LocationReviewResponse, status_code=201, summary="提交/更新地点评价",
             dependencies=[Depends(require_campus_verified())])
async def upsert_location_review(
    location_id: int,
    data: LocationReviewCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    tenant: TenantContext = Depends(get_tenant_context),
):
    """提交本人评价（每地点每用户一条，重复提交=更新）。"""
    lr = await db.execute(
        select(Location).where(Location.id == location_id, Location.is_deleted == False)
    )
    location = lr.scalar_one_or_none()
    if location is None:
        raise NotFoundException(detail="地点不存在")
    check_resource_in_tenant(location.school_id, tenant)

    # 已有评价则更新，否则新增
    rr = await db.execute(
        select(LocationReview).where(
            LocationReview.location_id == location_id,
            LocationReview.user_id == current_user.id,
            LocationReview.is_deleted == False,
        )
    )
    review = rr.scalar_one_or_none()
    if review is None:
        review = LocationReview(
            location_id=location_id,
            user_id=current_user.id,
            school_id=tenant.school_id,
            score=data.score,
            content=data.content,
            status="published",
        )
        db.add(review)
    else:
        review.score = data.score
        review.content = data.content

    await db.flush()
    await _recalc_location_rating(db, location)
    await mark_location_summary_dirty(db, location.id)
    await db.commit()

    # 返回最新评价（含作者）
    fresh = (await db.execute(
        select(LocationReview).options(joinedload(LocationReview.user))
        .where(LocationReview.id == review.id)
    )).scalar_one()
    return _review_response(fresh)


@router.delete("/locations/{location_id}/reviews", summary="撤回我的地点评价",
               dependencies=[Depends(require_campus_verified())])
async def delete_location_review(
    location_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    tenant: TenantContext = Depends(get_tenant_context),
):
    """撤回本人评价并重算评分。"""
    lr = await db.execute(
        select(Location).where(Location.id == location_id, Location.is_deleted == False)
    )
    location = lr.scalar_one_or_none()
    if location is None:
        raise NotFoundException(detail="地点不存在")
    check_resource_in_tenant(location.school_id, tenant)

    rr = await db.execute(
        select(LocationReview).where(
            LocationReview.location_id == location_id,
            LocationReview.user_id == current_user.id,
            LocationReview.is_deleted == False,
        )
    )
    review = rr.scalar_one_or_none()
    if review is not None:
        review.is_deleted = True
        review.deleted_at = datetime.now()
        await db.flush()
        await _recalc_location_rating(db, location)
        await mark_location_summary_dirty(db, location.id)
        await db.commit()

    return {"message": "撤回成功"}
